sjsonnet/src/sjsonnet/Std.scala

```diff
@@ -26,71 +27,416 @@ object Std {
+      val c = cStr.charAt(0)
+      val b = new mutable.ArrayBuilder.ofRef[Val.Lazy]
+      var i = 0
+      var start = 0
+      while(i < str.length) {
+        if(str.charAt(i) == c) {
+          val finalStr = Val.Str(pos, str.substring(start, i))
+          b.+=(new Val.Strict(finalStr))
+          start = i+1
+        }
+        i += 1
+      }
+      b.+=(new Val.Strict(Val.Str(pos, str.substring(start, math.min(i, str.length)))))
+      new Val.Arr(pos, b.result())
+    }
+  }
+
+  private object SplitLimit extends Val.Builtin3("str", "c", "maxSplits") {
+    def evalRhs(str: Val, c: Val, maxSplits: Val, ev: EvalScope, pos: Position): Val = {
+      new Val.Arr(pos, str.asString.split(java.util.regex.Pattern.quote(c.asString), maxSplits.asInt + 1).map(s => (new Val.Strict(Val.Str(pos, s)))))
+    }
+  }
+
+  private object StringChars extends Val.Builtin1("str") {
+    def evalRhs(str: Val, ev: EvalScope, pos: Position): Val =
+      stringChars(pos, str.asString)
+  }
+
+  private object ParseInt extends Val.Builtin1("str") {
+    def evalRhs(str: Val, ev: EvalScope, pos: Position): Val =
+      Val.Num(pos, str.asString.toInt)
+  }
+
+  private object ParseOctal extends Val.Builtin1("str") {
+    def evalRhs(str: Val, ev: EvalScope, pos: Position): Val =
+      Val.Num(pos, Integer.parseInt(str.asString, 8))
+  }
+
+  private object ParseHex extends Val.Builtin1("str") {
+    def evalRhs(str: Val, ev: EvalScope, pos: Position): Val =
+      Val.Num(pos, Integer.parseInt(str.asString, 16))
+  }
+
+  private object MD5 extends Val.Builtin1("s") {
+    def evalRhs(s: Val, ev: EvalScope, pos: Position): Val =
+      Val.Str(pos, Platform.md5(s.asString))
+  }
+
+  private object AsciiUpper extends Val.Builtin1("str") {
+    def evalRhs(str: Val, ev: EvalScope, pos: Position): Val =
+      Val.Str(pos, str.asString.toUpperCase)
+  }
+
+  private object AsciiLower extends Val.Builtin1("str") {
+    def evalRhs(str: Val, ev: EvalScope, pos: Position): Val =
+      Val.Str(pos, str.asString.toLowerCase)
+  }
+
+  private object Trace extends Val.Builtin2("str", "rest") {
+    def evalRhs(str: Val, rest: Val, ev: EvalScope, pos: Position): Val = {
+      System.err.println(s"TRACE: ${pos.fileScope.currentFileLastPathElement} " + str.asString)
+      rest
+    }
+  }
+
+  private object ExtVar extends Val.Builtin1("x") {
+    def evalRhs(_x: Val, ev: EvalScope, pos: Position): Val = {
+      val Val.Str(_, x) = _x
+      Materializer.reverse(
+        pos,
+        ev.extVars.getOrElse(x, throw new Error.Delegate("Unknown extVar: " + x))
+      )
+    }
+  }
+
+  val functions: Seq[(String, Val.Func)] = Seq(
+    "assertEqual" -> AssertEqual,
+    "toString" -> ToString,
+    "codepoint" -> Codepoint,
+    "length" -> Length,
+    "objectHas" -> ObjectHas,
+    "objectHasAll" -> ObjectHasAll,
+    "objectFields" -> ObjectFields,
+    "objectFieldsAll" -> ObjectFieldsAll,
+    builtin("objectValues", "o"){ (pos, ev, v1: Val.Obj) =>
```

Comment:
What's the distinction between the builtins you moved to `object`s, and those you left as-is using the `builtin` helper? Were the ones you rewrote the "hot" builtins, or was there some other distinguishing factor?

Reply:
Yes, I mainly rewrote the ones that are used in the benchmark. It's a lot of work...

Reply:
Got it, leaving them half-half is probably fine for now then